### PR TITLE
✨Source MailGun: Migrate Python CDK to Low-code CDK

### LINKED ISSUE #28113: Drift

### LINKED ISSUE #29120: mailgun

### PR DESCRIPTION
**Pull request by btkcodedev**
## What
*Migrating Source Mailgun to Low-Code CDK*
Closes #29120

## How
*Developed using (Configuration Based Source) low-code CDK*

## Recommended reading order
1. `spec.yaml`
2. `manifest.yaml`
3. `schemas/*`

## Tests

<details><summary><strong>Integration & Acceptance</strong></summary>
Full Test Results

</details>

## 🚨 User Impact 🚨
- No breaking changes, migration to low-code
<details><summary><strong>Migrate Connector: Source Mailgun</strong></summary>

### Community member or Airbyter

- [x] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [x] Code reviews completed
- [x] Documentation updated
    - [x] Connector's `README.md`
    - [x] Connector's `bootstrap.md`
    - [x] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [x] `docs/integrations/README.md`
    - [x] `airbyte-integrations/builds.md`
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

</details>

### CODE REVIEW

**Comment by btkcodedev** [comment]
Test results: Passed state:
image

**Comment by btkcodedev** [comment]
Waiting for #28471 for merging to update start_date

**Comment by marcosmarxm** [comment]
> Waiting for #28471 for merging to update start_date

Go ahead and fix the start date directly here.

File: airbyte-integrations/connectors/source-mailgun/acceptance-test-config.yml [hunk 1]
```diff
@@ -1,24 +1,37 @@
-# See [Connector Acceptance Tests](https://docs.airbyte.com/connector-development/testing-connectors/connector-acceptance-tests-reference)
-# for more information about how to configure these tests
 connector_image: airbyte/source-mailgun:dev
-tests:
+acceptance_tests:
   spec:
-    - spec_path: "source_mailgun/spec.json"
+    tests:
+      - spec_path: "source_mailgun/spec.yaml"
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-mailgun/acceptance-test-config.yml:5]: This is going to bypass the backward test.
```suggestion
      - spec_path: "source_mailgun/spec.yaml"
        backward_compatibility_tests_config:
          disable_for_version: 0.1.1

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-mailgun/metadata.yaml [hunk 2]
```diff
@@ -1,24 +1,25 @@
 data:
+  allowedHosts:
+    hosts:
+      - https://api.mailgun.net
+  registries:
+    oss:
+      enabled: false
+    cloud:
+      enabled: false
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-mailgun/metadata.yaml:9]: Why ar eyou changing this?

File: airbyte-integrations/connectors/source-mailgun/metadata.yaml [hunk 3]
```diff
@@ -1,24 +1,25 @@
...
   githubIssueLabel: source-mailgun
   icon: mailgun.svg
   license: MIT
   name: Mailgun
-  registries:
-    cloud:
-      enabled: true
-    oss:
-      enabled: true
+  releaseDate: TODO
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-mailgun/metadata.yaml:19]: Add a date here

File: airbyte-integrations/connectors/source-mailgun/source_mailgun/manifest.yaml [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+      type: "BasicHttpAuthenticator"
+      username: "api"
+      password: "{{ config['private_key'] }}"
+
+  retriever:
+    type: SimpleRetriever
+    record_selector:
+      $ref: "#/definitions/selector"
+    paginator:
+      type: NoPagination
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-mailgun/source_mailgun/manifest.yaml:29]: Please add pagination as it was implemented before https://github.com/airbytehq/airbyte/blob/b78ad4e9d905f2f02631f7897acabf6441375032/airbyte-integrations/connectors/source-mailgun/source_mailgun/source.py#L42-L49

File: airbyte-integrations/connectors/source-mailgun/source_mailgun/manifest.yaml [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+streams:
+  - "#/definitions/domains_stream"
+  - "#/definitions/events_stream"
+
+check:
+  type: CheckStream
+  stream_names:
+    - "domains"
+    - "events"
+
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-mailgun/source_mailgun/manifest.yaml:78]: EOF

**Review by marcosmarxm — DISMISSED** [review]
Left some comments, the connector is missing pagination.

**Comment by btkcodedev** [comment]
All comments resolved
Test result passed state

**Comment by btkcodedev** [comment]
image

**Comment by sajarin** [comment]
@btkcodedev getting an error with the check command with the events stream: 

`{"type": "LOG", "log": {"level": "ERROR", "message": "Encountered an error trying to connect to stream events. Error: \n Traceback (most recent call last):\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/declarative/checks/check_stream.py\", line 42, in check_connection\n    stream_is_available, reason = availability_strategy.check_availability(stream, logger, source)\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/streams/http/availability_strategy.py\", line 36, in check_availability\n    stream_slice = get_first_stream_slice(stream)\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/streams/utils/stream_helper.py\", line 21, in get_first_stream_slice\n    stream.stream_slices(\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/declarative/declarative_stream.py\", line 126, in stream_slices\n    return self.retriever.stream_slices()\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py\", line 348, in stream_slices\n    return self.stream_slicer.stream_slices()\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/declarative/incremental/datetime_based_cursor.py\", line 141, in stream_slices\n    start_datetime = self._calculate_earliest_possible_value(self._select_best_end_datetime())\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/declarative/incremental/datetime_based_cursor.py\", line 146, in _calculate_earliest_possible_value\n    earliest_possible_start_datetime = min(self.start_datetime.get_datetime(self.config), end_datetime)\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/declarative/datetime/min_max_datetime.py\", line 58, in get_datetime\n    time = self._parser.parse(str(self.datetime.eval(config, **additional_parameters)), datetime_format)\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/declarative/datetime/datetime_parser.py\", line 29, in parse\n    parsed_datetime = datetime.datetime.strptime(str(date), format)\n  File \"/usr/local/lib/python3.9/_strptime.py\", line 568, in _strptime_datetime\n    tt, fraction, gmtoff_fraction = _strptime(data_string, format)\n  File \"/usr/local/lib/python3.9/_strptime.py\", line 349, in _strptime\n    raise ValueError(\"time data %r does not match format %r\" %\nValueError: time data '' does not match format '%Y-%m-%dT%H:%M:%SZ'\n"}}`


`{"type": "LOG", "log": {"level": "ERROR", "message": "Check failed"}}`


`{"type": "CONNECTION_STATUS", "connectionStatus": {"status": "FAILED", "message": "\"Unable to connect to stream events - time data '' does not match format '%Y-%m-%dT%H:%M:%SZ'\""}}`

**Comment by btkcodedev** [comment]
@sajarin 
Please update the secrets config to match the updated format (start_date: %Y-%m-%dT%H:%M:%SZ)
Ping via Slack (btkcodedev) for getting my creds if needed

**Comment by sajarin** [comment]
@btkcodedev thanks for the update, config validation is now passing but no records are being fetched. I pinged you on slack for credentials so we can get this moving.

**Review by sajarin — APPROVED** [review]

**Comment by sajarin** [comment]
Thanks for the contribution @btkcodedev!

**Comment by padala** [comment]
@btkcodedev @sajarin Thanks for the contribution. I'm still facing the issue with sync resulting in 0 records when I tried this today on Airbyte Cloud. Is this already resolved? 

> @btkcodedev thanks for the update, config validation is now passing but no records are being fetched. I pinged you on slack for credentials so we can get this moving.

**Comment by sajarin** [comment]
@padala can you make a new issue and tag me? Perhaps an issue with the record selector

**Comment by padala** [comment]
@sajarin  I've created a new issues here - https://github.com/airbytehq/airbyte/issues/30105

**Pull request merged**